### PR TITLE
lgc: generalize traceNonUniformIndex

### PR DESCRIPTION
**Pull request by nhaehnle**
Clarify what the search of this function does: it scans through the data flow graph, trying to prove inputs to the original nonUniformVal uniform. It heuristically picks a GEP argument as a "remaining" non-uniform index.

Allow propagation through more instruction types in preparation for future changes.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9313682d2f7059fbaf58f7a9b974b212f6611f1e
<details><summary>CTS tests (Failed: 0/211876)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35496/70631 (50.3%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35135/70631 (49.7%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39379/70629 (55.8%)</li>
<li>Failed: 0/70629 (0.0%)</li>
<li>Not Supported: 31250/70629 (44.2%)</li><li>Warnings: 0/70629 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70616 (55.8%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31203/70616 (44.2%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1c6ed8aa0c0f04ac7ff439118d279deb2592d16d
<details><summary>CTS tests (Failed: 0/211876)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35496/70631 (50.3%)</li>
<li>Failed: 0/70631 (0.0%)</li>
<li>Not Supported: 35135/70631 (49.7%)</li><li>Warnings: 0/70631 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39379/70629 (55.8%)</li>
<li>Failed: 0/70629 (0.0%)</li>
<li>Not Supported: 31250/70629 (44.2%)</li><li>Warnings: 0/70629 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70616 (55.8%)</li>
<li>Failed: 0/70616 (0.0%)</li>
<li>Not Supported: 31203/70616 (44.2%)</li><li>Warnings: 0/70616 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
Ping for review

File: lgc/builder/BuilderImpl.cpp [hunk 1]
```diff
@@ -388,90 +380,133 @@ static Value *traceNonUniformIndex(Value *nonUniformVal) {
...
+        nonUniformValSize = size;
+      }
     }
-    if (auto gep = dyn_cast<GetElementPtrInst>(base)) {
-      if (gep->hasAllConstantIndices()) {
-        base = gep->getPointerOperand();
-        continue;
+
+    // See if we can propagate the search further.
+    auto propagate = [&](Value *value) -> bool {
```
Comment (perlfu) [inline lgc/builder/BuilderImpl.cpp:408]: Purely stylistic, but presumably declaration doesn't need to be in the loop?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7f7bc09598ec8e57d5f659ddd364c569dc14bad6
<details><summary>CTS tests (Failed: 0/211633)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35496/70549 (50.3%)</li>
<li>Failed: 0/70549 (0.0%)</li>
<li>Not Supported: 35053/70549 (49.7%)</li><li>Warnings: 0/70549 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39381/70550 (55.8%)</li>
<li>Failed: 0/70550 (0.0%)</li>
<li>Not Supported: 31169/70550 (44.2%)</li><li>Warnings: 0/70550 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70534 (55.9%)</li>
<li>Failed: 0/70534 (0.0%)</li>
<li>Not Supported: 31121/70534 (44.1%)</li><li>Warnings: 0/70534 (0.0%)</li>
</ul></details>


</ul></details>

**Review by perlfu — APPROVED** [review]
LGTM - as best as I can tell

It's a shame we cannot acquire the divergence/uniformity definitions of address spaces and intrinsics from backend definitions somehow, but that is no worse than the current code.

File: lgc/builder/BuilderImpl.cpp (hunk 1 above)
Comment (nhaehnle) [reply lgc/builder/BuilderImpl.cpp]: Thank you for taking a look. My thinking was to have it close to its uses, but I don't think it really matters much. I've moved it out of the loop.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 04e4db4dcebe01450f608dbc6d175c330ceb2eb4
<details><summary>CTS tests (Failed: 0/211630)</summary><ul><li>Built with version 1.3.3.1</li>
<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 35496/70549 (50.3%)</li>
<li>Failed: 0/70549 (0.0%)</li>
<li>Not Supported: 35053/70549 (49.7%)</li><li>Warnings: 0/70549 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39379/70547 (55.8%)</li>
<li>Failed: 0/70547 (0.0%)</li>
<li>Not Supported: 31168/70547 (44.2%)</li><li>Warnings: 0/70547 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39413/70534 (55.9%)</li>
<li>Failed: 0/70534 (0.0%)</li>
<li>Not Supported: 31121/70534 (44.1%)</li><li>Warnings: 0/70534 (0.0%)</li>
</ul></details>


</ul></details>

**Review by perlfu — APPROVED** [review]

**Comment by nhaehnle** [comment]
Thanks, Carl!

ping @amdrexu / @dstutt for review/approval

**Comment by nhaehnle** [comment]
Rebased and fixed a typo noticed by @jasilvanus on #2636.

**Review by amdrexu — APPROVED** [review]

**Pull request merged**